source/extensions/filters/network/mongo_proxy/codec_impl.cc

```diff
@@ -395,6 +524,43 @@ void EncoderImpl::encodeReply(const ReplyMessage& message) {
   }
 }
 
+void EncoderImpl::encodeCommand(const CommandMessage& message) {
+  int32_t total_size = 16;
+  total_size += message.database().size() + 1;
+  total_size += message.commandName().size() + 1;
```

Comment:
Same thing here - a bit unclear why some items get padding and others not.  Do you think it's worth a helper function for cStringSize() which would return string.size() + 1 with a comment in one place, or is that overkill?  This one optional since I'm only going to make you clean up so much preexisting code :-)

Reply:
@alyssawilk Thanks for reviewing the code. When I was working on that filter I noticed several things which bothered me, but wanted to wait until this PR is accepted before suggesting major changes. But since you pointed out several of those issues, maybe it is a good time to talk about them. As you noticed, adding support for new messages is copy&paste exercise. The code reuse is minimal. Basically, processing of each message is: restore from wire format, try to inject delay, increase counter and log the parsed message. That sequence is repeated for every single message (~6-8 times). The messages create sort of class hierarchy, but code never takes advantage of that hierarchy. I mean it does not operate on API defined by base class and handles differences via virtual methods. As you noticed there are lots of magic numbers (again - copy and paste). There are probably 3-4 other things I could mention but I stop here. I am wondering how open you are for restructuring of that code and make it more object oriented and eliminate code duplication? I have some ideas and even did quick prototype and it seems to work. The price would be longer code review, because the changes will not be incremental.

Reply:
I agree the code could probably be improved, but the flip side is this code hasn't changed in 2 years and it's extremely battle tested. We will also likely never add messages again as I think these are the remaining ones. My preference is to not do major restructuring unless there is a really good reason to do so. 

(Feel free to do cleanups to make things constants, etc. if you like, but would prefer to discuss larger changes beforehand).

Reply:
SGTM.  I wouldn't ask for major clean up as a part of this PR both because I'm not terribly familiar with Mongo (I had to do a bunch of googling around to even do the review I did) and also because I'm out a bunch this week (out tomorrow, and laggy the next 2 days).  Let's just go for minor things called out and if you're enthusiastic to do more clean up after we can always follow up with Matt.

Reply:
@mattklein123 Sure - I totally understand your reasoning. I will clean it up as per @alyssawilk code review. Version 3.6 of mongo https://docs.mongodb.com/manual/reference/mongodb-wire-protocol/#client-request-messages added new message OP_MSG (code 2013) for which we do not have support yet.